File: sdks/python/pulsely.py
"""
Pulsely server SDK for Python — standard library only.

    from pulsely import Pulsely

    bp = Pulsely(
        app_id=os.environ["PULSELY_APP_ID"],
        app_key=os.environ["PULSELY_APP_KEY"],
        app_secret=os.environ["PULSELY_APP_SECRET"],
        base_url="https://pulsely.example.com",
    )

    bp.trigger("orders", "created", {"id": 42})

The app secret must never reach a browser.
"""

import hashlib
import hmac
import json
import time
import urllib.error
import urllib.parse
import urllib.request


class PulselyError(Exception):
    """Raised when a publish is rejected or the request cannot be made."""

    def __init__(self, message, status=0, body=None):
        super().__init__(message)
        self.status = status
        self.body = body


class Pulsely:
    def __init__(self, app_id, app_secret, app_key=None,
                 base_url="http://127.0.0.1:8085", timeout=10):
        if not app_id or not app_secret:
            raise PulselyError("Pulsely: app_id and app_secret are required.")
        self.app_id = app_id
        self.app_key = app_key
        self.app_secret = app_secret
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout

    def trigger(self, channel, event, data=None):
        """Publish an event to a channel. Raises PulselyError on any non-2xx."""
        # Serialize once and send exactly these bytes. Re-serializing for the
        # request risks a different key order and a signature that no longer
        # matches the body — the most common cause of a puzzling 401.
        payload = json.dumps(
            {"channel": channel, "event": event, "data": data if data is not None else {}},
            separators=(",", ":"),
        ).encode("utf-8")

        path = f"/apps/{self.app_id}/events"
        timestamp = str(int(time.time()))  # epoch SECONDS, UTC

        request = urllib.request.Request(
            self.base_url + path,
            data=payload,
            method="POST",
            headers={
                "Content-Type": "application/json",
                "X-Pulsely-Timestamp": timestamp,
                "X-Pulsely-Signature": self._sign("POST", path, timestamp, payload),
            },
        )

        try:
            with urllib.request.urlopen(request, timeout=self.timeout) as response:
                return self._parse(response.read())
        except urllib.error.HTTPError as err:
            body = self._parse(err.read())
            message = body.get("error", body) if isinstance(body, dict) else body
            raise PulselyError(
                f"Pulsely: publish rejected ({err.code}) — {message}", err.code, body
            ) from None
        except urllib.error.URLError as err:
            raise PulselyError(f"Pulsely: request failed — {err.reason}") from None

    def list_channels(self, filter_by_prefix=None, info=None):
        """
        Every currently occupied channel, as {"channel_name": {}} — or
        {"user_count": N} for a presence channel when info="user_count".

        Raises PulselyError on any non-2xx.
        """
        path = f"/apps/{self.app_id}/channels"
        query = {}
        if filter_by_prefix:
            query["filter_by_prefix"] = filter_by_prefix
        if info:
            query["info"] = info
        return self._get(path, query).get("channels", {})

    def get_channel(self, channel_name):
        """
        Detail for one channel: {occupied, subscription_count, user_count?}.
        Returns rather than raising for an unoccupied channel — "not
        occupied" is a normal state, not a missing resource.

        Raises PulselyError on any non-2xx (e.g. an invalid channel name).
        """
        path = f"/apps/{self.app_id}/channels/{channel_name}"
        return self._get(path, {})

    def _get(self, path, query):
        """
        Signed GET request against the given (unsigned, no query string)
        path. Query params are appended after signing — they are never part
        of the signing string.
        """
        timestamp = str(int(time.time()))
        signature = self._sign("GET", path, timestamp, b"")
        url = self.base_url + path
        if query:
            url += "?" + urllib.parse.urlencode(query)

        request = urllib.request.Request(
            url,
            method="GET",
            headers={
                "X-Pulsely-Timestamp": timestamp,
                "X-Pulsely-Signature": signature,
            },
        )

        try:
            with urllib.request.urlopen(request, timeout=self.timeout) as response:
                return self._parse(response.read())
        except urllib.error.HTTPError as err:
            body = self._parse(err.read())
            message = body.get("error", body) if isinstance(body, dict) else body
            raise PulselyError(
                f"Pulsely: request rejected ({err.code}) — {message}", err.code, body
            ) from None
        except urllib.error.URLError as err:
            raise PulselyError(f"Pulsely: request failed — {err.reason}") from None

    def auth_token(self, user_id, ttl_seconds=3600):
        """
        Mint a short-lived connection token for a browser.

        The browser presents it as the STOMP passcode; it establishes the identity
        your auth endpoint later sees as `user_token`. Public channels need none.
        """
        if not self.app_key:
            raise PulselyError("Pulsely: app_key is required to mint connection tokens.")

        expires = int(time.time()) + ttl_seconds
        signature = hmac.new(
            self.app_secret.encode("utf-8"),
            f"{self.app_key}:{expires}:{user_id}".encode("utf-8"),
            hashlib.sha256,
        ).hexdigest()

        return f"{expires}.{user_id}.{signature}"

    def authorize_channel(self, authorized=True, user_id="", user_info=None):
        """
        Build the response your auth endpoint returns when Pulsely asks whether
        a subscriber may join a private or presence channel.

        Presence channels need user_id; user_info is echoed to every other member.
        """
        return {
            "authorized": bool(authorized),
            "user_id": str(user_id),
            "user_info": user_info if user_info is not None else {},
        }

    def verify_webhook(self, raw_body, headers=None, tolerance_seconds=300):
        """
        Verify an inbound event webhook really came from Pulsely.

        Pass the RAW request body bytes, not a re-serialized object.
        """
        headers = {k.lower(): v for k, v in (headers or {}).items()}
        timestamp = headers.get("x-pulsely-timestamp")
        signature = headers.get("x-pulsely-signature")
        if not timestamp or not signature:
            return False

        try:
            if abs(int(time.time()) - int(timestamp)) > tolerance_seconds:
                return False
        except ValueError:
            return False

        if isinstance(raw_body, str):
            raw_body = raw_body.encode("utf-8")

        expected = self._sign("POST", "/webhook", str(timestamp), raw_body)
        return hmac.compare_digest(expected, str(signature))

    def _sign(self, method, path, timestamp, payload):
        body_hash = hashlib.sha256(payload).hexdigest()
        # Joined by real newlines, not the two characters backslash-n.
        signing_string = f"{method.upper()}\n{path}\n{timestamp}\n{body_hash}"
        return hmac.new(
            self.app_secret.encode("utf-8"), signing_string.encode("utf-8"), hashlib.sha256
        ).hexdigest()

    @staticmethod
    def _parse(raw):
        text = raw.decode("utf-8", "replace")
        try:
            return json.loads(text)
        except ValueError:
            return text

File: sdks/node/pulsely.mjs
/**
 * Pulsely server SDK for Node.js — zero dependencies.
 *
 *   import Pulsely from './pulsely.mjs';
 *
 *   const bp = new Pulsely( {
 *     appId:     process.env.PULSELY_APP_ID,
 *     appKey:    process.env.PULSELY_APP_KEY,
 *     appSecret: process.env.PULSELY_APP_SECRET,
 *     baseUrl:   'https://pulsely.example.com'
 *   } );
 *
 *   await bp.trigger( 'orders', 'created', { id: 42 } );
 *
 * The app secret must never reach a browser.
 */
import crypto from 'node:crypto';

export class PulselyError extends Error {
	constructor( message, status, body ) {
		super( message );
		this.name = 'PulselyError';
		this.status = status;
		this.body = body;
	}
}

export default class Pulsely {

	constructor( { appId, appKey, appSecret, baseUrl = 'http://127.0.0.1:8085', timeoutMs = 10000 } = {} ) {
		if ( !appId || !appSecret ) {
			throw new PulselyError( 'Pulsely: appId and appSecret are required.' );
		}
		this.appId = appId;
		this.appKey = appKey;
		this.appSecret = appSecret;
		this.baseUrl = baseUrl.replace( /\/+$/, '' );
		this.timeoutMs = timeoutMs;
	}

	/**
	 * Publish an event to a channel.
	 *
	 * @throws {PulselyError} on any non-2xx, carrying status and parsed body.
	 */
	async trigger( channel, event, data = {} ) {
		// Serialize once and send exactly these bytes. Re-serializing for the
		// request risks a different key order and a signature that no longer
		// matches the body — the most common cause of a puzzling 401.
		const payload = JSON.stringify( { channel, event, data } );
		const path = `/apps/${this.appId}/events`;
		const timestamp = String( Math.floor( Date.now() / 1000 ) );  // epoch SECONDS, UTC

		const controller = new AbortController();
		const timer = setTimeout( () => controller.abort(), this.timeoutMs );

		let res;
		try {
			res = await fetch( this.baseUrl + path, {
				method: 'POST',
				headers: {
					'Content-Type': 'application/json',
					'X-Pulsely-Timestamp': timestamp,
					'X-Pulsely-Signature': this.#sign( 'POST', path, timestamp, payload )
				},
				body: payload,
				signal: controller.signal
			} );
		} catch ( err ) {
			throw new PulselyError( `Pulsely: request failed — ${err.message}`, 0, null );
		} finally {
			clearTimeout( timer );
		}

		const text = await res.text();
		let body;
		try { body = JSON.parse( text ); } catch { body = text; }

		if ( !res.ok ) {
			throw new PulselyError(
				`Pulsely: publish rejected (${res.status}) — ${body?.error ?? text}`,
				res.status, body
			);
		}
		return body;
	}

	/**
	 * Every currently occupied channel, as `{ channel_name: {} }` — or
	 * `{ user_count: N }` for a presence channel when `info: 'user_count'`.
	 *
	 * @throws {PulselyError} on any non-2xx.
	 */
	async listChannels( { filterByPrefix, info } = {} ) {
		const path = `/apps/${this.appId}/channels`;
		const query = {};
		if ( filterByPrefix ) query.filter_by_prefix = filterByPrefix;
		if ( info ) query.info = info;
		const body = await this.#get( path, query );
		return body.channels ?? {};
	}

	/**
	 * Detail for one channel: `{ occupied, subscription_count, user_count? }`.
	 * Resolves rather than rejecting for an unoccupied channel — "not
	 * occupied" is a normal state, not a missing resource.
	 *
	 * @throws {PulselyError} on any non-2xx (e.g. an invalid channel name).
	 */
	async getChannel( channelName ) {
		const path = `/apps/${this.appId}/channels/${channelName}`;
		return this.#get( path, {} );
	}

	/**
	 * Signed GET request against the given (unsigned, no query string) path.
	 * Query params are appended after signing — they are never part of the
	 * signing string.
	 */
	async #get( path, query ) {
		const timestamp = String( Math.floor( Date.now() / 1000 ) );
		const signature = this.#sign( 'GET', path, timestamp, '' );
		const qs = new URLSearchParams( query ).toString();
		const url = this.baseUrl + path + ( qs ? `?${qs}` : '' );

		const controller = new AbortController();
		const timer = setTimeout( () => controller.abort(), this.timeoutMs );

		let res;
		try {
			res = await fetch( url, {
				headers: {
					'X-Pulsely-Timestamp': timestamp,
					'X-Pulsely-Signature': signature
				},
				signal: controller.signal
			} );
		} catch ( err ) {
			throw new PulselyError( `Pulsely: request failed — ${err.message}`, 0, null );
		} finally {
			clearTimeout( timer );
		}

		const text = await res.text();
		let body;
		try { body = JSON.parse( text ); } catch { body = text; }

		if ( !res.ok ) {
			throw new PulselyError(
				`Pulsely: request rejected (${res.status}) — ${body?.error ?? text}`,
				res.status, body
			);
		}
		return body;
	}

	/**
	 * Mint a short-lived connection token for a browser.
	 *
	 * The browser presents this as the STOMP passcode; it establishes the identity
	 * your auth endpoint later sees as `user_token`. Public channels need no token.
	 */
	authToken( userId, ttlSeconds = 3600 ) {
		if ( !this.appKey ) {
			throw new PulselyError( 'Pulsely: appKey is required to mint connection tokens.' );
		}
		const expires = Math.floor( Date.now() / 1000 ) + ttlSeconds;
		const signature = crypto
			.createHmac( 'sha256', this.appSecret )
			.update( `${this.appKey}:${expires}:${userId}` )
			.digest( 'hex' );

		return `${expires}.${userId}.${signature}`;
	}

	/**
	 * Build the response your auth endpoint returns when Pulsely asks whether a
	 * subscriber may join a private or presence channel.
	 *
	 * Presence channels need userId; user_info is echoed to every other member.
	 */
	authorizeChannel( { authorized = true, userId = '', userInfo = {} } = {} ) {
		return { authorized, user_id: String( userId ), user_info: userInfo };
	}

	/**
	 * Verify an inbound event webhook really came from Pulsely.
	 *
	 *   if ( !bp.verifyWebhook( rawBody, req.headers ) ) return res.sendStatus( 401 );
	 *
	 * Pass the RAW request body, not a re-serialized object.
	 */
	verifyWebhook( rawBody, headers = {}, toleranceSeconds = 300 ) {
		const timestamp = headers[ 'x-pulsely-timestamp' ] ?? headers[ 'X-Pulsely-Timestamp' ];
		const signature = headers[ 'x-pulsely-signature' ] ?? headers[ 'X-Pulsely-Signature' ];
		if ( !timestamp || !signature ) return false;

		if ( Math.abs( Math.floor( Date.now() / 1000 ) - Number( timestamp ) ) > toleranceSeconds ) {
			return false;
		}

		const expected = this.#sign( 'POST', '/webhook', String( timestamp ), rawBody );
		const a = Buffer.from( expected );
		const b = Buffer.from( String( signature ) );
		return a.length === b.length && crypto.timingSafeEqual( a, b );
	}

	#sign( method, path, timestamp, payload ) {
		const bodyHash = crypto.createHash( 'sha256' ).update( payload ).digest( 'hex' );
		// Joined by real newlines, not the two characters backslash-n.
		const signingString = `${method.toUpperCase()}\n${path}\n${timestamp}\n${bodyHash}`;
		return crypto.createHmac( 'sha256', this.appSecret ).update( signingString ).digest( 'hex' );
	}

}
